design/pv-cloning.md

```diff
@@ -0,0 +1,54 @@
+# Cloning PVs While Remapping Namespaces
+
+Status: Draft
+
+Velero supports restoring resources into different namespaces than they were backed up from.
+This enables a user to, among other things, clone a namespace within a cluster.
+However, if the namespace being cloned uses persistent volume claims, Velero cannot currently create a second copy of the original persistent volume when restoring.
+This limitation is documented in detail in [issue #192](https://github.com/heptio/velero/issues/192).
+This document proposes a solution that allows new copies of persistent volumes to be created during a namespace clone.
+
+## Goals
+
+- Enable persistent volumes to be cloned when using `velero restore create --namespace-mappings ...` to create a second copy of a namespace within a cluster.
+
+## Non Goals
+
+- Cloning of persistent volumes in any scenario other than when using `velero restore create --namespace-mappings ...` flag.
+- [CSI-based cloning](https://kubernetes.io/docs/concepts/storage/volume-pvc-datasource/).
+
+## Background
+
+(Omitted, see introduction)
+
+## High-Level Design
+
+During a restore, Velero will detect that it needs to assign a new name to a persistent volume being restored if and only if both of the following conditions are met:
```

Comment:
To make sure I understand how a user would do the clone operation, the workflow would look something like this:

1. User does `velero backup create --include-namespaces ns1 src`
1. User then does `velero restore create --from-backup src --namespace-mappings ns1=ns2`

Is that right?

How will this interact with resource selection? If we just target PVs will it work?

Reply:
Yeah, that's what the workflow would look like. If you just do `--include-resources pvs`, then the `--namespace-mappings` flag doesn't really make sense since you're not restoring any namespaced resources; in any case, the design as proposed would not create a second copy of the PV in that scenario.

Reply:
Got it, that makes sense.

So if the NS already exists, it would behave like a normal Velero restore - anything already existing wouldn't be overridden. I know that's another issue, so I'm fine with this design not addressing it explicitly.

How about interaction with restic? Would this work to create a PVR and copy from the restic repository?

Reply:
Correct, yeah this won't do a delete/recreate if other resources already exist.

With restic, we actually avoid this issue entirely, because we rely on a dynamic re-provisioning of an empty PV to be able to do a restic restore into, and the dynamic provisioner will come up with a new PV name. So that should already work today.